silx/io/h5py_utils.py

```diff
@@ -0,0 +1,328 @@
+
+    :param str filename:
+    :yields File:
+    """
+    with File(filename) as h5file:
+        yield h5file
+
+
+@retry_contextmanager()
+def open_item(filename, name, retry_invalid=False, validate=None):
+    """Yield an HDF5 dataset or group (retry until it can be instantiated).
+
+    :param str filename:
+    :param bool retry_invalid: retry when item is missing or not valid
+    :param callable or None validate:
+    :yields Dataset, Group or None:
+    """
+    with File(filename) as h5file:
+        try:
+            item = h5file[name]
+        except KeyError as e:
+            if "doesn't exist" in str(e):
+                if retry_invalid:
+                    raise retry_mod.RetryError
+                else:
+                    item = None
+            else:
+                raise
+        if callable(validate) and item is not None:
+            if not validate(item):
+                if retry_invalid:
+                    raise retry_mod.RetryError
+                else:
+                    item = None
+        yield item
+
+
+def _top_level_names(filename, include_only=include_default):
+    """Return all valid top-level HDF5 names.
+
+    :param str filename:
+    :param callable or None include_only:
+    :returns list(str):
+    """
+    with File(filename) as h5file:
+        try:
+            if callable(include_only):
+                return [name for name in h5file["/"] if include_only(h5file[name])]
+            else:
+                return list(h5file["/"])
+        except KeyError:
+            raise retry_mod.RetryError
+
+
+top_level_names = retry()(_top_level_names)
+safe_top_level_names = retry_in_subprocess()(_top_level_names)
+
+
+class File(h5py.File):
+    """Takes care of HDF5 file locking and SWMR mode without the need
+    to handle those explicitely.
+
+    When using this class, you cannot open different files simultatiously
+    with different modes because the locking flag is an environment variable.
+    """
+
+    _HDF5_FILE_LOCKING = None
+    _NOPEN = 0
+    _SWMR_LIBVER = "latest"
+
+    def __init__(
+        self,
+        filename,
+        mode=None,
+        enable_file_locking=None,
+        swmr=None,
+        libver=None,
+        **kwargs
+    ):
+        """The arguments `enable_file_locking` and `swmr` should not be
+        specified explicitly for normal use cases.
+
+        :param str filename:
+        :param str or None mode: read-only by default
+        :param bool or None enable_file_locking: by default it is disabled for `mode='r'`
+                                                 and `swmr=False` and enabled for all
+                                                 other modes.
+        :param bool or None swmr: try both modes when `mode='r'` and `swmr=None`
+        :param **kwargs: see `h5py.File.__init__`
+        """
+        if mode is None:
+            mode = "r"
+        elif mode not in ("r", "w", "w-", "x", "a"):
```

Comment:
'r+' is missing. By the way according to what you discovered with the 'r' mode (and probably r+ ?) maybe it is worth adding a warning if libver == 110 ?

Reply:
A python warning or a warning in the docs? If a python warning: a warning on the writer side saying "file cannot be read until closed" or a warning on the reader side saying "will fail to open when being written"?

Reply:
A python warning. My thought was more or on the reader side.

Reply:
```python
# writer
import h5py
f=h5py.File("test.h5", mode="a", libver="latest")
f.flush()
```

```python
# reader
import h5py
f=h5py.File("test.h5", mode="r")

OSError: Unable to open file (file is already open for write (may use <h5clear file> to clear file consistency flags))
```

If you want a warning on the reader side, you'll have to check the libver the file is written in. This means you need to open the file which mean you already get the exception "file is already open for write". In any case, I don't think it is useful the have a warning just before the exception basically saying the same thing.

Reply:
:ok_hand: